pkg/engine/datasource/kafka_datasource/sarama_config_parameters_test.go

```diff
@@ -323,25 +323,39 @@ func testAsyncProducer(t *testing.T, options *GraphQLSubscriptionOptions, start,
 	asyncProducer, err := sarama.NewAsyncProducer(options.BrokerAddresses, config)
 	require.NoError(t, err)
 
-	for i := start; i < end; i++ {
-		message := &sarama.ProducerMessage{
-			Topic: options.Topic,
-			Value: sarama.StringEncoder(fmt.Sprintf(messageTemplate, i)),
+	for _, topic := range options.Topics {
+		values, ok := messages[topic]
+		if ok {
+			for _, value := range values {
+				message := &sarama.ProducerMessage{
+					Topic: topic,
+					Value: sarama.StringEncoder(value),
+				}
+				asyncProducer.Input() <- message
+			}
 		}
-		asyncProducer.Input() <- message
 	}
 }
 
-func testConsumeMessages(messages chan *sarama.ConsumerMessage, numberOfMessages int) (map[string]struct{}, error) {
-	allMessages := make(map[string]struct{})
+func consumeTestMessages(t *testing.T, messages chan *sarama.ConsumerMessage, producedMessages map[string][]string) {
+	var expectedNumMessages int
+	for _, values := range producedMessages {
+		expectedNumMessages += len(values)
+	}
+
+	consumedMessages := make(map[string][]string)
+	var numMessages int
 	for msg := range messages {
+		numMessages++
+		topic := msg.Topic
 		value := string(msg.Value)
-		allMessages[value] = struct{}{}
-		if len(allMessages) >= numberOfMessages {
-			return allMessages, nil
+		consumedMessages[topic] = append(consumedMessages[topic], value)
+		if numMessages >= expectedNumMessages {
```

Comment:
Are too many messages not a problem? It would always break when equal, but what if there were many more to come?

Reply:
The number of messages is carefully chosen for every test. A Kafka topic is an unbounded stream of messages. I set a hard time limit to wait for messages. Now the consumer quits when all produced messages are consumed or it fails with an error message if the produced messages are not consumed in 10 seconds. 

In our test suite, a fixed number of messages are produced first, then the consumer consumes all messages immediately and compares them with the produced messages.